File: day3.go
package advent2023

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
	"testing"
)

// GetLinesFromFile reads a file and returns its lines as a slice of strings.
func GetLinesFromFile(filename string) ([]string, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	return lines, scanner.Err()
}
func TestDay3Task1(t *testing.T) {
	lines, err := GetLinesFromFile("day3.txt")
	if err != nil {
		t.Fatalf("Error reading file: %v", err)
	}

	var allNumbers []*PosNumber
	for lineIndex, line := range lines {
		numbersForLine := findNumbers(line)

		fmt.Printf("%s\n", line)

		for _, res := range numbersForLine {
			// record the lineIndex for found numbers
			res.LineIndex = lineIndex
			fmt.Printf("line: %d, cols:(%d-%d), n:%d\n ", res.LineIndex, res.StartPos, res.EndPos, res.Number)
		}
		allNumbers = append(allNumbers, numbersForLine...)
	}

	// Now we have all numbers and their locations in allNumbers.
	// When we inspect the surrounding for symbols we should
	// be able to find out what numbers are valid
	// to use those in the totalSum

	totalSum := 0
	for _, number := range allNumbers {
		if symbolAround(number, lines) {
			totalSum += number.Number
		}
	}

	fmt.Printf("totalSum: %d\n", totalSum)
}

func symbolAround(number *PosNumber, lines []string) bool {
	from := number.StartPos - 1
	if from < 0 {
		from = 0
	}
	to := number.EndPos + 1
	if to > len(lines[0]) {
		to = len(lines[0])
	} // assume all lines have same len

	// loop three lines
	for looplines := number.LineIndex - 1; looplines <= number.LineIndex+1; looplines++ {
		if looplines < 0 || looplines >= len(lines) {
			continue
		}
		// inspect line characters
		symbolFound := strings.IndexAny(lines[looplines][from:to], "+#$*@/=%-&")
		// we know enough already
		if symbolFound > -1 {
			return true
		}
	}

	return false
}

type PosNumber struct {
	StartPos  int
	EndPos    int
	LineIndex int
	Number    int
}

func findNumbers(str string) []*PosNumber {
	// Define the regular expression for finding numbers
	re := regexp.MustCompile(`\d+`)

	// Find all matches in the input string
	matches := re.FindAllStringSubmatchIndex(str, -1)

	// Extract numbers along with start and end positions
	result := make([]*PosNumber, len(matches))
	for i, match := range matches {
		start := match[0]
		end := match[1]
		number, _ := strconv.Atoi(str[start:end])

		result[i] = &PosNumber{
			Number:   number,
			StartPos: start,
			EndPos:   end,
		}
	}

	return result
}

// We have to find numbers that are adjacent to the gear(*) symbol,
// but only the gears that have two adjacent numbers are important.
// Not sure if we might have gears with 3 numbers around it btw..
func TestDay3Task2(t *testing.T) {
	lines, err := GetLinesFromFile("day3.txt")
	if err != nil {
		t.Fatalf("Error reading file: %v", err)
	}
	var allNumbers []*PosNumber
	for lineIndex, line := range lines {
		numbersForLine := findNumbers(line)

		fmt.Printf("%s\n", line)

		for _, res := range numbersForLine {
			// record the lineIndex for found numbers
			res.LineIndex = lineIndex
			//fmt.Printf("line: %d, cols:(%d-%d), n:%d\n ", res.LineIndex, res.StartPos, res.EndPos, res.Number)
		}
		allNumbers = append(allNumbers, numbersForLine...)
	}

	// Now we have all numbers and their locations in allNumbers.

	totalSum := 0

	// now we just loop over all gear(*) symbols to search
	// for surrounding numbers on adjacent lines
	for lineIndexofStar, l := range lines {
		fmt.Printf("%s\n", l)
		starPositions := findCharPositions(l, '*')

		for _, starPos := range starPositions {
			var foundNumbers []*PosNumber
			for _, number := range allNumbers {
				if numberTouchesSymbol(number, lineIndexofStar, starPos, len(lines)) {
					foundNumbers = append(foundNumbers, number)
				}
			}
			// if the foundnumbers length is two, we have to add to the total
			if len(foundNumbers) == 2 {
				totalSum += foundNumbers[0].Number * foundNumbers[1].Number
			}
		}

	}

	fmt.Printf("totalSum: %d\n", totalSum)
}

func numberTouchesSymbol(number *PosNumber, lineIndex, starPos, numberoflines int) bool {
	// loop three lines
	for looplines := lineIndex - 1; looplines <= lineIndex+1; looplines++ {
		if number.LineIndex != looplines {
			continue
		}
		if looplines < 0 || looplines > numberoflines {
			continue
		}
		// inspect if this number touches the star
		if number.StartPos == starPos || number.StartPos == starPos+1 {
			return true // startposition of number is ok
		}
		if number.EndPos == starPos || number.EndPos-1 == starPos {
			return true
		}
		// when the lineIndex of number is not same, just check the boundarie
		if number.LineIndex != lineIndex {
			if number.StartPos <= starPos && number.EndPos >= starPos {
				return true
			}
		}
	}

	return false
}

func findCharPositions(str string, targetChar rune) []int {
	positions := []int{}

	for i, char := range str {
		if char == targetChar {
			positions = append(positions, i)
		}
	}

	return positions
}

File: day3.py
import re

class PosNumber:
    def __init__(self, start_pos, end_pos, line_index, number):
        self.start_pos = start_pos
        self.end_pos = end_pos
        self.line_index = line_index
        self.number = number

def get_lines_from_file(filename):
    with open(filename, 'r') as file:
        return file.readlines()

def find_numbers(line):
    pattern = re.compile(r'\d+')
    matches = pattern.finditer(line)
    return [PosNumber(match.start(), match.end(), 0, int(match.group())) for match in matches]

def symbol_around(number, lines):
    from_index = max(0, number.start_pos - 1)
    to = min(len(lines[0]), number.end_pos + 1)
    for looplines in range(max(0, number.line_index - 1), min(len(lines), number.line_index + 2)):
        if any(c in lines[looplines][from_index:to] for c in "+#$*@/=%-&"):
            return True
    return False

def test_day3_task1():
    lines = get_lines_from_file("day3-test.txt")
    all_numbers = []
    for line_index, line in enumerate(lines):
        numbers_for_line = find_numbers(line)
        for number in numbers_for_line:
            number.line_index = line_index
        all_numbers.extend(numbers_for_line)

    total_sum = sum(number.number for number in all_numbers if symbol_around(number, lines))
    print("totalSum:", total_sum)

def number_touches_symbol(number, line_index, star_pos, number_of_lines):
    for looplines in range(line_index - 1, line_index + 2):
        if looplines < 0 or looplines >= number_of_lines or number.line_index != looplines:
            continue
        if number.start_pos in [star_pos, star_pos + 1] or number.end_pos in [star_pos, star_pos - 1]:
            return True
        if number.line_index != line_index and number.start_pos <= star_pos <= number.end_pos:
            return True
    return False

def find_char_positions(line, target_char):
    return [i for i, char in enumerate(line) if char == target_char]

def test_day3_task2():
    lines = get_lines_from_file("day3-test.txt")
    all_numbers = []
    for line_index, line in enumerate(lines):
        numbers_for_line = find_numbers(line)
        for number in numbers_for_line:
            number.line_index = line_index
        all_numbers.extend(numbers_for_line)

    total_sum = 0
    for line_index, line in enumerate(lines):
        star_positions = find_char_positions(line, '*')
        for star_pos in star_positions:
            found_numbers = [number for number in all_numbers if number_touches_symbol(number, line_index, star_pos, len(lines))]
            #print(":",found_numbers)
            if len(found_numbers) == 2:
                total_sum += found_numbers[0].number * found_numbers[1].number
                print("Found:", found_numbers[0].number, " x ", found_numbers[1].number)

    print("totalSum:", total_sum)

# To run the tests, uncomment the following lines:
# test_day3_task1()
test_day3_task2()
